experimental/packages/otlp-transformer/src/metrics/index.ts

```diff
@@ -0,0 +1,103 @@
+import type { MetricRecord } from '@opentelemetry/sdk-metrics-base';
+import { toAttributes } from '../common/internal';
+import { toMetric } from './internal';
+import type { IExportMetricsServiceRequest } from './types';
+
+export function createExportMetricsServiceRequest(metricRecords: MetricRecord[], startTime: number): IExportMetricsServiceRequest | null {
+  if (metricRecords.length === 0) {
+    return null;
+  }
+
+  return {
+    resourceMetrics: metricRecordsToResourceMetrics(metricRecords).map(({ resource, resourceMetrics, schemaUrl: resourceSchemaUrl }) => ({
+      resource: {
+        attributes: toAttributes(resource.attributes),
+        droppedAttributesCount: 0,
+      },
+      instrumentationLibraryMetrics: resourceMetrics.map(({ instrumentationLibrary, instrumentationLibraryMetrics, schemaUrl: librarySchemaUrl }) => ({
+        instrumentationLibrary: {
+          name: instrumentationLibrary.name,
+          version: instrumentationLibrary.version,
+        },
+        metrics: instrumentationLibraryMetrics.map(m => toMetric(m, startTime)),
+        schemaUrl: librarySchemaUrl,
+      })),
+      schemaUrl: resourceSchemaUrl,
+    }))
+  };
+}
+
+type ResourceMetrics = {
+  resource: Resource,
+  resourceMetrics: InstrumentationLibraryMetrics[],
+  schemaUrl?: string,
+};
+
+type InstrumentationLibraryMetrics = {
+  instrumentationLibrary: InstrumentationLibrary,
+  instrumentationLibraryMetrics: MetricRecord[],
+  schemaUrl?: string,
+};
+
+function metricRecordsToResourceMetrics(metricRecords: MetricRecord[]): ResourceMetrics[] {
+  const resourceMap: Map<Resource, Map<string, MetricRecord[]>> = new Map();
+
+  for (const record of metricRecords) {
+    let ilmMap = resourceMap.get(record.resource);
+
+    if (!ilmMap) {
+      ilmMap = new Map();
+      resourceMap.set(record.resource, ilmMap);
+    }
+
+    const instrumentationLibraryKey = `${record.instrumentationLibrary.name}@${record.instrumentationLibrary.name || ''}:${record.instrumentationLibrary.schemaUrl || ''}`;
+    let records = ilmMap.get(instrumentationLibraryKey);
+
+    if (!records) {
+      records = [];
+      ilmMap.set(instrumentationLibraryKey, records);
+    }
+
+    records.push(record);
+  }
+
+  const out: ResourceMetrics[] = [];
+
+  const resourceMapEntryIterator = resourceMap.entries();
+  let resourceMapEntry = resourceMapEntryIterator.next();
+  while (!resourceMapEntry.done) {
+    const [resource, ilmMap] = resourceMapEntry.value;
+    const resourceMetrics: InstrumentationLibraryMetrics[] = [];
+    const ilmIterator = ilmMap.values();
+    let ilmEntry = ilmIterator.next();
+    while (!ilmEntry.done) {
+      const instrumentationLibraryMetrics = ilmEntry.value;
+      if (instrumentationLibraryMetrics.length > 0) {
+        const lib = instrumentationLibraryMetrics[0].instrumentationLibrary;
+        resourceMetrics.push({ instrumentationLibrary: lib, instrumentationLibraryMetrics, schemaUrl: lib.schemaUrl });
```

Comment:
Noticed 1 thing - the instrumentation library schema URL is converted to the resource schema URL. Is it supposed to be like this? See the comment here: https://github.com/open-telemetry/opentelemetry-proto/blob/v0.14.0/opentelemetry/proto/metrics/v1/metrics.proto#L55-L58

Just need clarification, noticed while prepping a followup to this PR to use the new SDK metric types and can rectify it there.

Reply:
good catch

Reply:
actually it doesn't work like you're thinking. The resource schema url would live next to the resourcemetrics array, not within it. In any case, the types in this particular function are not otlp types so I renamed them to make it more obvious what they represent.